compiler/rustc_target/src/spec/apple_base.rs

```diff
@@ -1,8 +1,46 @@
 use std::{borrow::Cow, env};
 
-use crate::spec::{cvs, FramePointer, LldFlavor, SplitDebuginfo, TargetOptions};
+use crate::spec::{
+    cvs, FramePointer, LinkArgs, LinkerFlavor, LldFlavor, SplitDebuginfo, TargetOptions,
+};
+
+fn pre_link_args(os: &'static str, arch: &'static str, target_abi: Option<&str>) -> LinkArgs {
+    let mut args = LinkArgs::new();
+
+    let platform_name = match target_abi {
+        Some("sim") => format!("{}-simulator", os),
+        Some("macabi") => "mac-catalyst".to_string(),
+        _ => os.to_string(),
+    };
+
+    let platform_version = match os.as_ref() {
+        "ios" => ios_lld_platform_version(),
+        "tvos" => tvos_lld_platform_version(),
+        "watchos" => watchos_lld_platform_version(),
+        "macos" => macos_lld_platform_version(arch),
+        _ => unreachable!(),
+    };
+
+    if target_abi != Some("macabi") {
+        args.insert(LinkerFlavor::Gcc, vec!["-arch".into(), arch.into()]);
+    }
+
+    args.insert(
+        LinkerFlavor::Lld(LldFlavor::Ld64),
+        vec![
+            "-arch".into(),
+            arch.into(),
+            "-platform_version".into(),
+            platform_name.into(),
+            platform_version.clone().into(),
+            platform_version.into(),
+        ],
+    );
+
+    args
+}
 
-pub fn opts(os: &'static str) -> TargetOptions {
+pub fn opts(os: &'static str, arch: &'static str, target_abi: Option<&str>) -> TargetOptions {
```

Comment:
```suggestion
pub fn opts(os: &'static str, arch: &'static str, abi: &'static str) -> TargetOptions {
```
The ABI passed to here is sometimes `None` and sometimes `Some("")`, but both are used interchangeably, so it probably makes sense to eliminate the `Option`.
(Also the `target_` prefix is kind of a tautology here.)